jwst/badpix_selfcal/badpix_selfcal_step.py

```diff
@@ -0,0 +1,188 @@
+        i.e., true self-calibration.
+        """
+
+        input_sci, selfcal_list, bkg_list_asn = _parse_inputs(input, selfcal_list)
+        # ensure that there are background exposures to use, otherwise skip the step
+        # unless forced
+        if (len(selfcal_list) == 0) and (not self.force_single):
+            log.warning("No selfcal or background exposures provided for self-calibration. \
+                        Skipping step. If you want to force self-calibration with the science \
+                        exposure alone (generally not recommended), set force_single=True.")
+            input_sci.meta.cal_step.badpix_selfcal = 'SKIPPED'
+            return input_sci, bkg_list_asn
+
+        # get the dispersion axis
+        try:
+            dispaxis = input_sci.meta.wcsinfo.dispersion_direction
+        except AttributeError:
+            log.warning("Dispersion axis not found in input science image metadata.\
+                        Kernel for self-calibration will be two-dimensional.")
+            dispaxis = None
+
+        # collapse all selfcal exposures into a single background model
+        # note that selfcal_list includes the science exposure. This is expected.
+        # all exposures are combined into a single background model using a MIN operation.
+        selfcal_list = [input_sci] + [dm.open(k) for k in selfcal_list]
+        selfcal_3d = []
+        for i, selfcal_model in enumerate(selfcal_list):
+            selfcal_3d.append(selfcal_model.data)
+        minimg = np.nanmin(np.asarray(selfcal_3d), axis=0)
+        bad_indices = badpix_selfcal.badpix_selfcal(minimg, self.flagfrac_lower, self.flagfrac_upper, self.kernel_size, dispaxis)
+
+        # apply the flags to the science data
+        input_sci = badpix_selfcal.apply_flags(input_sci, bad_indices)
+
+        # apply the flags to the background data to be passed to background sub step
+        if len(bkg_list_asn) > 0:
+            for i, background_model in enumerate(bkg_list_asn):
+                bkg_list_asn[i] = badpix_selfcal.apply_flags(dm.open(background_model), bad_indices)
+
+        input_sci.meta.cal_step.badpix_selfcal = 'COMPLETE'
+        return input_sci, bkg_list_asn
+
+
+def _parse_inputs(input, selfcal_list):
+    """
+    Parse the input to the step. This is a helper function that is used in the
+    command line interface to the step.
+
+    Parameters
+    ----------
+    input: str
+        input file or association, or tuple of (sci, bkg, selfcal)
+
+    selfcal_list: list of user-supplied ImageModels or filenames to use for selfcal
+
+    Returns
+    -------
+    input: JWST data model or association
+        input science data to be corrected
+
+    selfcal_list: list of ImageModels or filenames to use for selfcal
+    """
+    selfcal_list_user = selfcal_list
+
+    if isinstance(input, tuple):
+        input_sci = dm.open(input[0])
+        selfcal_list = list(input[1]) + list(input[2])
+        bkg_list_asn = input[1]
```

Comment:
This works with the way it is called in spec2 now, but it is a little convoluted.  Why not just have the step accept a selfcal_list, used only for selfcal, and a bkg_list with background files to update?  That way a user directly calling the step could also distinguish between selfcal files and background files, without having to make an association file, or know that they should put them in a tuple in the input variable.

Reply:
this is a great suggestion.  In working on this, I also realized it might be more intuitive for users if, in the case that an .asn file with selfcal exposures AND selfcal_list is not None, the selfcal_list were appended to the association selfcal list (rather than having the user input override it).  What do you think of that?

Reply:
That sounds useful to me!